cpp/src/plasma/store.cc

```diff
@@ -73,19 +64,68 @@ using arrow::cuda::CudaDeviceManager;
 using arrow::util::ArrowLog;
 using arrow::util::ArrowLogLevel;
 
-namespace fb = plasma::flatbuf;
-
 namespace plasma {
 
+using flatbuf::MessageType;
+
 void SetMallocGranularity(int value);
 
 struct GetRequest {
-  GetRequest(Client* client, const std::vector<ObjectID>& object_ids);
+  GetRequest(asio::io_context& io_context,
+             const std::shared_ptr<ClientConnection>& client,
+             const std::vector<ObjectID>& object_ids)
+      : client(client),
+        object_ids(object_ids.begin(), object_ids.end()),
+        objects(object_ids.size()),
+        num_satisfied(0),
+        timer_(io_context) {
+    std::unordered_set<ObjectID> unique_ids(object_ids.begin(), object_ids.end());
+    num_objects_to_wait_for = unique_ids.size();
+  }
+
+  void ReturnFromGet() {
+    // Figure out how many file descriptors we need to send.
+    std::unordered_set<int> fds_to_send;
+    std::vector<int> store_fds;
+    std::vector<int64_t> mmap_sizes;
+    for (const auto& object_id : object_ids) {
+      PlasmaObject& object = objects[object_id];
+      int fd = object.store_fd;
+      if (object.data_size != -1 && fds_to_send.count(fd) == 0 && fd != -1) {
+        fds_to_send.insert(fd);
+        store_fds.push_back(fd);
+        mmap_sizes.push_back(GetMmapSize(fd));
+      }
+    }
+
+    // Send the get reply to the client.
+    Status s = SendGetReply(client, &object_ids[0], objects, object_ids.size(), store_fds,
```

Comment:
Nit: this *looks* like it could segfault if object_ids is empty. Could be made more clear with:
```suggestion
    Status s = SendGetReply(client, object_ids.data(), objects, object_ids.size(), store_fds,
```